### ISSUE TITLE
Maintenance migrations doesn't get executed when registering migrations via .ScanIn(typeof(Program).Assembly).For.Migrations()

### ISSUE THREAD

**Issue by ModernRonin** [fluentmigrator/fluentmigrator#1607]
**Describe the bug**

Maintenance migrations don't seem to be executed.

**To Reproduce**
I have a maintenance migration defined as 

```csharp
[Maintenance(MigrationStage.BeforeAll, TransactionBehavior.None)]
[UsedImplicitly]
public class EnsureMigrationsCannotBeRunSimultaneouslyByMultipleInstances : Migration
{
    public override void Down() => throw new NotImplementedException("Down migrations are not supported for sp_getapplock");

    public override void Up()
    {
        throw new Exception("x");
        Console.WriteLine("START!!!!!!!!!!!!!!!!!!!!!!!!!!");
        Execute.Sql($@"
            DECLARE @result INT
            EXEC @result = sp_getapplock '{GetType().Namespace}', 'Exclusive', 'Session'

            IF @result < 0
            BEGIN
                DECLARE @msg NVARCHAR(1000) = 'Received error code ' + CAST(@result AS VARCHAR(10)) + ' from sp_getapplock during migrations';
                THROW 99999, @msg, 1;
            END
        ");
    }
}
```
Note that I throw an exception in `Up()`.

When I run the migrations in a unit-test, they all run through just fine and *I never hit the exception*.
In fact, all the structure described in my migrations is created happily.

So I am under the impression this specific 'before-all' migration is never executed.

**Expected behavior**
The migration is executed and thus the exception thrown.

**Information (please complete the following information):**
 - OS: Windows 11
 - Platform: Net6
 - FluentMigrator version: 3.3.1
 - FluentMigrator runner: "in-process runner" 
 - Database Management System  SQLServer (docker image mcr.microsoft.com/mssql/server:2019-CU5-ubuntu-18.04)
 - Database Management System Version: see the answer to the last question, sqlserver 2019-CU5

**Additional context**
Basically, I've been following the guidance from [here](https://fluentmigrator.github.io/articles/faq.html#how-can-i-run-migrations-safely-from-multiple-application-servers) to prevent multiple service instances running migrations at the same time. 

However, I just got round to check whether this actually works and it seems these special maintenance migrations are never called.

I tried to step into the code of FluentMigrator, as much as VS and the nugets let me, and in `MigrationRunner.MigrateUp()` where it calls `ApplyMaintenance(MigrationStage.BeforeAll, useAutomaticTransactionManagement);` which leads to `MaintenanceLoader.LoadMaintenance` - and in there `_maintenace` is empty. Dunno if that helps.

**Comment by jzabroski** [comment]
If you're using the in-process runner, are you setting up the scan call for the maintenance migrations.  Maintenance migrations are currently not assumed to be in the same assembly as your regulation migrations.

Related, we would love a PR for #1231

**Comment by ModernRonin** [comment]
Regarding the first question, I use the following extension method:
```csharp
public static IServiceCollection AddMigrations(this IServiceCollection self, string connectionString, Type initialMigrationType)
    {
        self.AddFluentMigratorCore()
            .ConfigureRunner(rb => rb
                .AddSqlServer()
                .WithMigrationsIn()
                .WithGlobalConnectionString(connectionString)
                .ScanIn(initialMigrationType.Assembly, typeof(Cleanup).Assembly)
                .For.Migrations());
        return self;
    }
```

The second assembly passed (typeofCleanup) is the one containing the maintenance migrations. Or am I doing something wrong here?

Regarding the second question, all projects reference the same version (we use [paket](https://fsprojects.github.io/Paket/index.html) which enforces this for all nugets).

**Comment by jzabroski** [comment]
See: https://github.com/fluentmigrator/fluentmigrator/issues/1062#issuecomment-616598419 - I think this is actually the heart of it.

Nobody ever did a PR to fix this/smooth it over, but we did get to the bottom of it / provide workarounds.

**Comment by ModernRonin** [comment]
Sorry, it took me a while to try this out, swamped with other stuff.

Removing WithMigrationsIn() does not change any behavior - so, you're right it's unnecessary, however, I still don't hit the exception.

The workaround works fine, though :-)

Thank you, not just for the help, but generally for this awesome library!

**Comment by jzabroski** [comment]
This will be fixed in 5.0.0 with a breaking change.

**Comment by nathanpovo** [comment]
Still an issue in version `5.2.0`.

---

As an alternative to `.ScanIn(XYZ).For.All()`, I created an extension method:

```csharp
public static IScanInBuilder RegisterAssemblySourceItem(this IScanInBuilder builder)
{
    builder.Services.AddSingleton<IAssemblySourceItem>(builder.SourceItem);

    return builder;
}
```

to allow for more granualirity when scanning in assemblies.

Example usage (modified code from the sample):

```csharp
private static void RunWithServices(DatabaseConfiguration dbConfig)
{
    // Initialize the services
    var serviceProvider = new ServiceCollection()
        .AddLogging(lb => lb.AddDebug().AddFluentMigratorConsole())
        .AddFluentMigratorCore()
        .ConfigureRunner(
            builder => builder
                .AddSqlServer()
                .AddSQLite()
                .WithGlobalConnectionString(dbConfig.ConnectionString)
                .ScanIn(typeof(AddGTDTables).Assembly)
                    .For.Migrations()
                    .For.EmbeddedResources()
                    .RegisterAssemblySourceItem() // Register the assembly source item
                .ScanIn(typeof(AddGTDTables).Assembly)
                    .For.Migrations()
                    .For.EmbeddedResources()
                    .For.VersionTableMetaData()
                    .RegisterAssemblySourceItem() // Register the assembly source item
            )
        .Configure<SelectingProcessorAccessorOptions>(
            opt => opt.ProcessorId = dbConfig.ProcessorId)
        .BuildServiceProvider();

    // Instantiate the runner
    var runner = serviceProvider.GetRequiredService<IMigrationRunner>();

    // Run the migrations
    runner.MigrateUp();
}
```

**Comment by perlun** [comment]
Still an issue with 7.0.0. Changing to `.For.All()` seems to do the trick nicely. 🎉 

@jzabroski, the comment you added in d0c4c362138bd4bb60cdc0d317236bcf2d54ded5 would be nice get published on the web site as well. 🤔 As it is right now, the approach described on [in-process migration runners](https://fluentmigrator.github.io/articles/migration-runners.html?tabs=vs-pkg-manager-console) leads you "wrong" if you want to use maintenance migrations, and the section about them [here](https://fluentmigrator.github.io/articles/faq.html#how-can-i-run-migrations-safely-from-multiple-application-servers) also doesn't (from what I can tell) mention anything about this important aspect of getting them working.

**Comment by jzabroski** [comment]
makes sense. updated the docs.

**Comment by jzabroski** [comment]
Your suggestion is published here: https://fluentmigrator.github.io/articles/faq.html#why-arent-my-maintenance-migrations-found-and-run-by-my-in-process-runner

**Comment by perlun** [comment]
Thanks @jzabroski, appreciate the effort. Would it even make sense to update the example at https://fluentmigrator.github.io/articles/migration-runners.html?tabs=vs-pkg-manager-console to do the `For.All()` approach rather than `For.Migrations()`, or does it come with any significant disadvantages?

**Comment by jzabroski** [comment]
Created the PR here https://github.com/fluentmigrator/documentation/pull/81
